pkg/handler/data_record_frame.go

```diff
@@ -0,0 +1,92 @@
+package handler
+
+import (
+	"encoding/base64"
+	"encoding/json"
+
+	"github.com/brandur/simplebox"
+	"github.com/checkr/flagr/pkg/util"
+	"github.com/checkr/flagr/swagger_gen/models"
+)
+
+type dataRecordEncryptor interface {
+	Encrypt([]byte) (string, error)
+}
+
+type simpleboxEncryptor struct{ key [simplebox.KeySize]byte }
+
+func (se *simpleboxEncryptor) Encrypt(b []byte) (string, error) {
+	s := base64.StdEncoding.EncodeToString(
+		simplebox.NewFromSecretKey(&se.key).Encrypt(b),
+	)
+	return s, nil
+}
+
+func newSimpleboxEncryptor(k string) dataRecordEncryptor {
+	key := [simplebox.KeySize]byte{}
+	copy(key[:], k)
+	return &simpleboxEncryptor{key: key}
+}
+
+const (
+	frameOutputModePayloadRawJSON = "payload_raw_json"
+)
+
+// DataRecordFrameOptions represents the options we can set to create a DataRecordFrame
+type DataRecordFrameOptions struct {
+	Encrypted       bool
+	Encryptor       dataRecordEncryptor
+	FrameOutputMode string
+}
+
+// DataRecordFrame represents the structure we can json.Marshal into data recorders
+type DataRecordFrame struct {
+	Encrypted      bool            `json:"encrypted"`
+	Payload        string          `json:"payload"`
```

Comment:
I think that at the very least we should add `omitempty` to both `Payload` and `PayloadRawJSON` as there is only ever going to be one or the other. The more I've though about this however I've come to think that it would be nice to achieve a couple of different goals:

1) Keep the output JSON message as `payload` whether or not options are set to use raw payload or not. In each case, the resulting data is the payload.
2) Forgo the "encrypted" field if we have the raw JSON setting turned on. If we are outputting raw JSON then the result will by definition not be encrypted.

With these ideas in mind, I've created an alternative version of the this file that takes these scenarios into account. I'm sure there is room for some cleanup but I'd like to see what you think of this idea:

```go
package handler

import (
	"encoding/base64"
	"encoding/json"

	"github.com/brandur/simplebox"
	"github.com/checkr/flagr/pkg/util"
	"github.com/checkr/flagr/swagger_gen/models"
)

type dataRecordEncryptor interface {
	Encrypt([]byte) (string, error)
}

type simpleboxEncryptor struct{ key [simplebox.KeySize]byte }

func (se *simpleboxEncryptor) Encrypt(b []byte) (string, error) {
	s := base64.StdEncoding.EncodeToString(
		simplebox.NewFromSecretKey(&se.key).Encrypt(b),
	)
	return s, nil
}

func newSimpleboxEncryptor(k string) dataRecordEncryptor {
	key := [simplebox.KeySize]byte{}
	copy(key[:], k)
	return &simpleboxEncryptor{key: key}
}

const (
	frameOutputModePayloadRawJSON = "payload_raw_json"
)

// DataRecordFrameOptions represents the options we can set to create a DataRecordFrame
type DataRecordFrameOptions struct {
	Encrypted       bool
	Encryptor       dataRecordEncryptor
	FrameOutputMode string
}

type rawPayload struct {
	Payload json.RawMessage `json:"payload"`
}

type stringPayload struct {
	Payload   string `json:"payload"`
	Encrypted bool
}

// DataRecordFrame represents the structure we can json.Marshal into data recorders
type DataRecordFrame struct {
	evalResult models.EvalResult
	options    DataRecordFrameOptions
}

func (drf *DataRecordFrame) MarshalJSON() ([]byte, error) {
	payload, err := drf.evalResult.MarshalBinary()
	if err != nil {
		return nil, err
	}

	if drf.options.FrameOutputMode == frameOutputModePayloadRawJSON {
		return json.Marshal(&rawPayload{
			Payload: payload,
		})
	}

	if drf.options.Encrypted && drf.options.Encryptor != nil {
		encryptedPayload, err := drf.options.Encryptor.Encrypt(payload)
		if err != nil {
			return nil, err
		}
		return json.Marshal(&stringPayload{
			Payload:   encryptedPayload,
			Encrypted: true,
		})
	}

	return json.Marshal(&stringPayload{
		Payload:   string(payload),
		Encrypted: false,
	})
}

// GetPartitionKey gets the partition key from entityID
func (drf *DataRecordFrame) GetPartitionKey() string {
	if drf.evalResult.EvalContext == nil {
		return ""
	}
	return util.SafeString(drf.evalResult.EvalContext.EntityID)
}

// Output sets the paylaod using its input and returns the json marshal bytes
func (drf *DataRecordFrame) Output() ([]byte, error) {
	return json.Marshal(drf)
}

```

Reply:
I like it, much better!